clifford/__init__.py

```diff
@@ -155,87 +157,101 @@ def construct_tables(
-                m = m_list_copy[ind]
-                l = l_list_copy[ind]
-                output[l] += value[k] * mult_table_vals_copy[ind] * other_value[m]
-            return output
-
-        return mv_mult
-
-    elif ((grades_a is not None) and (grades_b is not None)):
-        # We can also specify sparseness by grade
-        filter_mask = np.zeros(len(k_list_copy), dtype=bool)
-        for i in range(len(filter_mask)):
-            if gradeList[k_list_copy[i]] in grades_a:
-                if gradeList[m_list_copy[i]] in grades_b:
-                    filter_mask[i] = 1
+    Returns
+    -------
+    func : function (array_like (n_dims,), array_like (n_dims,)) -> array_like (n_dims,)
+        A function that computes the appropriate multiplication
+    """
+    @numba.njit
+    def mv_mult(value, other_value):
+        ret_dtype = (value[:0] * other_value[:0] * mult_table_vals[:0]).dtype
+        output = np.zeros(n_dims, dtype=ret_dtype)
+        for k, l, m, val in zip(k_list, l_list, m_list, mult_table_vals):
+            output[l] += value[k] * val * other_value[m]
+        return output
 
-        k_list_copy = k_list_copy[filter_mask]
-        l_list_copy = l_list_copy[filter_mask]
-        m_list_copy = m_list_copy[filter_mask]
-        mult_table_vals_copy = mult_table_vals_copy[filter_mask]
+    return mv_mult
 
-        @numba.njit
-        def mv_mult(value, other_value):
-            output = np.zeros(n_dims)
-            for ind, k in enumerate(k_list_copy):
-                m = m_list_copy[ind]
-                l = l_list_copy[ind]
-                output[l] += value[k] * mult_table_vals_copy[ind] * other_value[m]
-            return output
 
-        return mv_mult
+def _get_mult_function_runtime_sparse(k_list, l_list, m_list, mult_table_vals, n_dims):
+    """
+    Get a multiplication function that attempts to exploit runtime zeros
+
+    See `_get_mult_function` for parameter descriptions.
+
+    The returned function avoids performing multiplications if vectors contain
+    zeros.
 
-    # This case we specify no sparseness in advance, the algorithm checks for zeros
+    TODO: determine if this actually helps.
+    """
     @numba.njit
     def mv_mult(value, other_value):
-        output = np.zeros(n_dims)
-        for ind, k in enumerate(k_list_copy):
+        ret_dtype = (value[:0] * other_value[:0] * mult_table_vals[:0]).dtype
+        output = np.zeros(n_dims, dtype=ret_dtype)
+        for ind, k in enumerate(k_list):
             v_val = value[k]
             if v_val != 0.0:
-                m = m_list_copy[ind]
+                m = m_list[ind]
                 ov_val = other_value[m]
                 if ov_val != 0.0:
-                    l = l_list_copy[ind]
-                    output[l] += v_val * mult_table_vals_copy[ind] * ov_val
+                    l = l_list[ind]
+                    output[l] += v_val * mult_table_vals[ind] * ov_val
         return output
```

Comment:
@hugohadfield: Does this actually help vs just doing the multiply by zero?

If not, this entire function can be removed.

Reply:
The first check almost definitely should (if v_val != 0.0:) but the second im not so sure. Once upon a time I did make a notebook comparing these things.. I think it is worth setting up a proper benchmarking system